src/docker/remote.rs

```diff
@@ -0,0 +1,653 @@
+                Err(_) => return true,
+            };
+            let file_name = entry.file_name();
+            !(file_type.is_file() || file_name == "src" || file_name == "etc")
+        },
+    )?;
+    copy_volume_files(engine, container, temppath, &dst.join("lib"), verbose)?;
+    // must make the `dst.join("lib")` **after** here, or we copy temp into lib.
+    create_volume_dir(engine, container, &dst_rustlib, verbose)?;
+
+    // we first copy over the toolchain file, then everything besides it.
+    // since we don't want to call docker 100x, we copy the intermediate
+    // files to a temp directory so they're cleaned up afterwards.
+    let toolchain_path = src_rustlib.join(&target.triple());
+    if toolchain_path.exists() {
+        copy_volume_files(engine, container, &toolchain_path, &dst_rustlib, verbose)?;
+    }
+
+    // now we need to copy over the host toolchain too, since it has
+    // some requirements to find std libraries, etc.
+    let rustc = sysroot.join("bin").join("rustc");
+    let libdir = Command::new(rustc)
+        .args(&["--print", "target-libdir"])
+        .run_and_get_stdout(verbose)?;
+    let host_toolchain_path = Path::new(libdir.trim()).parent().unwrap();
+    copy_volume_files(
+        engine,
+        container,
+        host_toolchain_path,
+        &dst_rustlib,
+        verbose,
+    )?;
+
+    Ok(())
+}
+
+pub fn volume_create(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "volume")
+        .args(&["create", volume])
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn volume_rm(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "volume")
+        .args(&["rm", volume])
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn volume_exists(engine: &Engine, volume: &str, verbose: bool) -> Result<bool> {
+    subcommand(engine, "volume")
+        .args(&["inspect", volume])
+        .run_and_get_output(verbose)
+        .map(|output| output.status.success())
+        .map_err(Into::into)
+}
+
+pub fn container_stop(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "stop")
+        .arg(container)
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn container_rm(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    subcommand(engine, "rm")
+        .arg(container)
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn container_state(engine: &Engine, container: &str, verbose: bool) -> Result<ContainerState> {
+    let stdout = subcommand(engine, "ps")
+        .arg("-a")
+        .args(&["--filter", &format!("name={container}")])
+        .args(&["--format", "{{.State}}"])
+        .run_and_get_stdout(verbose)?;
+    ContainerState::new(stdout.trim())
+}
+
+fn path_hash(path: &Path) -> String {
```

Comment:
this fn should be used in `Dockerfile::image_name` too

Reply:
Going to wait on #826 before changing this.

Reply:
`path_hash` has been moved to `docker::shared`, and is used in `image_name` and `unique_toolchain_identifier` (which is used in `unique_container_identifier`.